[Qi Ting] iP

nqt230 opened the pull request:
# Duke prototype 0.2
> "By failing to prepare, you are preparing to fail."
> – Benjamin Franklin


**Features**
- ~Command Line~ Text-based with GUI interface
- _Super_ fast
- Easy to use 🤔
    - Straightforward text commands to perform all functions
- Task managing functionality:
    - Supports creation of multiple kinds of tasks
    - Marking of tasks
    - Deletion of tasks
    - Local saving and loading of tasks


**Setting up:**
1. Download the latest release from [here](https://github.com/nqt230/ip/releases).
2. Run the file (double-click duke.jar).
3. Done! You can start using Duke to manage your tasks.


If you're a Java programmer, you can use it to practice Java too. Here's the `main` method:
```
public class Launcher {
    public static void main(String[] args) {
        Application.launch(Main.class, args);
    }
}
```


**Planned features**
- [X] Basic create, update and delete operations for task management.
- [ ] Task filtering and sorting by date.
- [ ] Task tagging.

Commit:
Add Gradle support

nqt230 committed:
Implement Level-1

- Add basic greeting and exit phrases
- Implement basic input reading

nqt230 committed:
Implement Level-2

- Create ParseInput method to parse commands to the bot and call the relevant methods
- Create AddToList and DisplayList methods to support text list feature
> AddToList stores a given text string to a list
> DisplayList prints all the text strings stored so far in order
- Remove unused Echo method from Level-1

nqt230 committed:
Implement Level-3

- Rework text list feature to a to-do list
> Create Task class to represent tasks that can be added to the to-do list
> Implement methods to support marking of Tasks as completed
- Implement ParseArgs method to support chatbot commands with arguments

nqt230 committed:
Implement Level-4

- Create Deadline, Event and ToDo classes that inherit from Task to support different types of Task creation
- Refactor input parsing logic to support creation of different types of tasks

nqt230 committed:
Implement A-TextUiTesting

- Add test input to input.txt file
- Performed testing using Runtest.bat
- Fix bug where tasks could be created with empty name or date

nqt230 committed:
Implement Level-5

- Create DukeException class to be thrown by and handled by methods in Duke
- Refactor methods to use DukeException class in error handling

nqt230 committed:
Implement Level-6

- Implement delete command
> Tasks can now be deleted from the list by index
- Modify test input to test deletion feature

nqt230 committed:
Implement Level-7

- Chat history with Duke is now saved between runs and loaded on start

nqt230 committed:
Implement Level-8

- Modify parsing of input to parse date arguments
- Modify Deadline and Event class to store date as LocalDate instead of String

nqt230 committed:
Merge branch 'branch-Level-8'

nqt230 committed:
Clean up and refactor code

- Modify all code to follow coding style
- Refactor task saving logic to be cleaner
- Update javadocs

nqt230 committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

nqt230 committed:
Implement A-Gradle

- Set up gradle

nqt230 committed:
Implement A-MoreOOP

- Create Ui, TaskList, Storage, and Parser classes to handle different operations of Duke
- Refactor code to these new classes

nqt230 committed:
Implement A-Packages

- Packaged all classes into one package Duke

nqt230 committed:
Implement A-JUnit

- Added JUnit tests for adding tasks, loading and saving tasks, and parsing input commands

nqt230 committed:
Implement A-Jar

nqt230 committed:
Implement A-JavaDoc

- Added JavaDocs to more methods

nqt230 committed:
Implement A-CodingStandard

- Adjust code to follow coding standard where not done previously

nqt230 committed:
Implement Level-9

- Add getter for taskName to Task class
- Add search task by taskName functionality to TaskList class
- Modify Parser class to accept search command
- Add display text function for search command to Ui class

nqt230 committed:
Merge branch 'branch-A-CodingStandard'

nqt230 committed:
Merge branch 'branch-Level-9'

nqt230 committed:
Add Javadocs headers to Level-9 methods

nqt230 committed:
Implement A-Checkstyle

- Add checkstyle plugin to gradle
- Use chcekstyle to find coding style violations
- Resolve coding style violations

nqt230 committed:
Implement Level-10

- Create GUI for Duke
    - User input is now received from the TextField component in MainWindow
    - handleUserInput function in MainWindow passes the user input to Duke
    - Duke takes the input and passes it to Parser
    - Duke now also stores the response to the user generated by Ui to be displayed by MainWindow

nqt230 committed:
Edit JavaDocs header

RachelChua commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,231 @@
+    public static Task addEvent(String args) throws DukeException {
+        String[] argsArr = args.split(" /at ", 2);
+        if (argsArr.length < 2) {
+            throw new DukeException("Failed to create event: Invalid number of arguments");
```
I like how you handle many different types of errors. Good job on error handling!

nqt230 replied:
Thank you!

RachelChua commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,231 @@
+package duke;
```
Perhaps you can divide into multiple packages 😃

nqt230 replied:
Haha will take note

RachelChua commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,63 @@
+ * A class used to represent a task. A task has a name and completion status.
+ */
+public abstract class Task {
+    protected static final String MARK_CHARACTER = "X";
```
Good use of static constants!

RachelChua commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,136 @@
+/**
+ * A class that handles the displaying of information to the user.
+ */
+public class Ui {
```
I like the abstraction and OOP used here!

RachelChua reviewed (commented):
LGTM! I like your error handling and OOP used! Also, your code follows the java coding standards very well. Good job!

totsukatomofumi commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,54 @@
+            break;
+        default:
+            throw new DukeException("Command not recognised");
+        }
```
I like how simple yet powerful you have made your parser.

nqt230 replied:
Thank you!

totsukatomofumi commented on src/main/java/duke/ToDo.java:
```diff
@@ -0,0 +1,20 @@
+/**
+ * Class used to represent a ToDo type task that has no date.
+ */
+public class ToDo extends Task {
```
I like the use of inheritance for a more OOP approach.

totsukatomofumi commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,63 @@
+     *
+     * @return A string that can be used to reconstruct the task.
+     */
+    public abstract String toSaveFormatString();
```
I like the naming idea for this method.

totsukatomofumi commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,136 @@
+            + "| | | | | | | |/ / _ \\\n"
+            + "| |_| | |_| |   <  __/\n"
+            + "|____/ \\__,_|_|\\_\\___|\n";
+    private static final String BOT_NAME = "Duke";
```
I like the use of a constant for this string that will be used often.

totsukatomofumi reviewed (commented):
LGTM! Very simple yet beautiful logic of code. Coding standard has been well adhered to. Good job!

nqt230 committed:
Implement A-Assertions

- Add Java asserts to handle some previously unhandled exceptions
    - TaskList: Use asserts to ensure null tasks cannot be added to the list
    - Duke: Use asserts to ensure Duke initializes correctly

nqt230 committed:
Improve code quality

* Refactor application launch process.
    - Move Launcher.main method to Duke.main.
    - Remove Launcher class since it is no longer used.
* Add constant TASK_TYPE_CHARACTER to ToDo, Deadline and Event. This is to centralize the information of which character refers to which type of task e.g. T associated to ToDo.
    - Modify toString and toSaveFormatString methods in all subclasses of Task to make use of this constant.
    - Modify Storage class to check against this constant when attempting to load in tasks from file.
* Move print statements to Duke's addToResponse method since all text to be displayed goes through that method.

nqt230 committed:
Merge pull request #2 from nqt230/branch-A-CodeQuality

Improve code quality

nqt230 committed:
Merge branch 'master' into branch-A-Assertions

nqt230 committed:
Merge pull request #3 from nqt230/branch-A-Assertions

Branch a assertions

nqt230 committed:
Update Duke.main javadocs header

nqt230 committed:
Implement A-CI

nqt230 committed:
Fix CI not set up correctly

- Move gradle.yml file to .github\workflows folder

nqt230 committed:
Implement C-Sort

- Add sorting command to Parser
- Add functionality to TaskList to sort by name
- Update Ui to respond correctly to new command

nqt230 committed:
Merge pull request #4 from nqt230/BCD-Extensions

Implement C-Sort

nqt230 committed:
Modify event and deadline creation in tasklist

- Adjust argument format to be more flexible

nqt230 committed:
Update tests and adjust exception message

- Update JUnit tests to account for changes to command format
- Remove new line from TaskList::addEvent exception message for invalid arguments

nqt230 committed:
Improve GUI and user friendliness

- Change GUI color scheme to dark mode
- Add help command to display available commands
- Add command format info to user feedback when user inputs invalid command

nqt230 committed:
Merge pull request #5 from nqt230/branch-A-BetterGUI

Improve GUI and user friendliness

nqt230 committed:
Add Ui.png and edit User Guide

nqt230 committed:
Update User Guide

- Add usage information for all commands to user guide

nqt230 committed:
Merge pull request #6 from nqt230/A-UserGuide

Implement A-UserGuide

nqt230 committed:
Replace Ui.png with better app screenshot
